File: src/main/java/cat/inspiracio/calculator/FzWorld.java
/*	Copyright 2011 Alexander Bunkenburg alex@inspiracio.cat
 * 
 * This file is part of Complex Calculator.
 * 
 * Complex Calculator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Complex Calculator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Complex Calculator. If not, see <http://www.gnu.org/licenses/>.
 * */
package cat.inspiracio.calculator;

import cat.inspiracio.complex.Complex;
import cat.inspiracio.complex.Complex$;
import cat.inspiracio.numbers.*;
import cat.inspiracio.parsing.SyntaxTree;

import java.awt.event.*;

import static cat.inspiracio.calculator.Interaction.MOVE;

final class FzWorld extends World {

    // connections -----------------------------

    private ZWorld zW;

	//State --------------------------------------------------------------------
	
    private ECList current;
    private PicletList piclets;
    private SyntaxTree f;

	//Constructor --------------------------------------------------------------

    FzWorld(Calculator calculator){
        super(calculator);
        setTitle("f(z) World");

        // GUI --------------------

        super.interaction = MOVE;

        MouseAdapter mouse = new MouseAdapter() {

            @Override public void mousePressed(MouseEvent mouseevent){
                prevx = mouseevent.getX();
                prevy = mouseevent.getY();
                mouseevent.consume();
            }

            @Override public void mouseReleased(MouseEvent mouseevent){
                int i = mouseevent.getX();
                int j = mouseevent.getY();
                canvas.shift(prevx - i, prevy - j);
                canvas.paint(canvas.getGraphics());
                prevx = i;
                prevy = j;
                mouseevent.consume();
            }
        };

        MouseMotionAdapter motion = new MouseMotionAdapter() {

            @Override public void mouseDragged(MouseEvent mouseevent){
                int i = mouseevent.getX();
                int j = mouseevent.getY();
                canvas.shift(prevx - i, prevy - j);
                canvas.paint(canvas.getGraphics());
                prevx = i;
                prevy = j;
                mouseevent.consume();
            }

        };

        plane.addMouseListener(mouse);
        plane.addMouseMotionListener(motion);
        sphere.addMouseListener(mouse);
        sphere.addMouseMotionListener(motion);
        pack();
        setLocationRelativeTo(calculator);
        setVisible(true);
    }

    private void setLocation(){
        setLocationRelativeTo(this.calculator);
    }

    //Methods ------------------------------------------------------------
    
    @Override void add(Complex c){
        if(f != null){
            try{
                Complex z = f.evaluate(c);
                current = new ECList(z, current);
                updateExtremes(z);
            }
            catch(Exception _ex) { }
            canvas.paint(canvas.getGraphics());
        }
    }

    void add(Piclet piclet){
        if(f != null){
            resetArg();
            ECList samples = piclet.getSamples();
            current = null;
            for(; samples != null; samples = samples.tail() )
                try{
                    Complex z = samples.head();
                    Complex c = f.evaluate(z);
                    updateExtremes(c);
                    current = new ECList(c, current);
                }
                catch(Exception _ex) { }
            stopDynamicMap();
            super.canvas.repaint();
        }
    }

    void add(PicletList ps){
        for(; ps != null; ps = ps.tail())
            add(ps.head());
    }

    void addCurrent(Piclet piclet){
        if(f != null){
            resetArg();
            current = null;
            for(ECList samples = piclet.getSamples(); samples != null; samples = samples.tail())
                try{
                    Complex z = samples.head();
                    Complex c = f.evaluate(z);
                    current = new ECList(c, current);
                }
                catch(Exception _ex) { }
            canvas.paint(canvas.getGraphics());
        }
    }

    private void resetArg(){Complex$.MODULE$.resetArg();}

    final void drawStuff(Drawing drawing){
        if(current != null)
            canvas.drawECList(drawing, current);
        for(PicletList ps = piclets; ps != null; ps = ps.tail())
            canvas.drawPiclet(drawing, ps.head());
    }

    void erase(){
        current = null;
        piclets = null;
        resetExtremes();
        canvas.repaint();
    }

    void functionChange(SyntaxTree syntaxtree){
        f = syntaxtree;
        erase();
        add(zW.getPiclets());
    }

    void setzWorld(ZWorld zworld){zW = zworld;}

    void stopDynamicMap(){
        piclets = new PicletList(new Freeline(current), piclets);
        current = null;
    }

}

File: src/main/scala/cat/inspiracio/calculator/FzWorld.scala
/*	Copyright 2011 Alexander Bunkenburg alex@inspiracio.cat
 *
 * This file is part of Complex Calculator.
 *
 * Complex Calculator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Complex Calculator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Complex Calculator. If not, see <http://www.gnu.org/licenses/>.
 * */
package cat.inspiracio.calculator

import java.awt.{Graphics, Point}

import cat.inspiracio.calculator.Interaction.MOVE
import cat.inspiracio.complex.Complex
import cat.inspiracio.geometry.{Curve, Piclet, Point2}

final class FzWorld private[calculator](override val calculator: Calculator) extends World(calculator) {
  import calculator.{f, zW}

  //State --------------------------------------------------------------------

  /** In z->f(z) mode, while the user drags on the z world, this is the curve
    * of resulting f(z). Otherwise null.
    * (That procedure is called "dynamic map".)
    * Can zs be improved somehow? Make it more local? */
  private var zs: List[Complex] = null

  /** The piclets that are displayed on the f(z) world. Usually curves. */
  private var piclets: List[Piclet] = Nil

  // initialisation -----------------------------------------------------------

  init()

  private def init() = {
    import icon._
    title = "f(z)"

    toolbar.peer.addSeparator()

    //move button always selected: just there to show user that they can move
    val moveButton = toggle(handIcon, "Move")
    moveButton.selected = true
    moveButton.enabled = false
    toolbar.contents += moveButton

    interaction = MOVE
    pack()
    applyPreferences()
    visible = true
  }

      /** start and previous mouse position during dragging */
      private var startPoint: Point = null
      private var previousPoint: Point = null

      override def mousePressed(point: Point) = {
        startPoint = point
        previousPoint = point
        canvas.startShift(startPoint)
      }

      override def mouseDragged(point: Point) = {
        canvas.shift(startPoint, previousPoint, point)
        previousPoint = point
      }

      override def mouseReleased(point: Point) = {
        canvas.endShift(startPoint, point)
      }

  private def applyPreferences() = {
    val p = preferences

    // to the right of z-world
    val zWorldDimension = zW.size //550 372
    val zWorldPosition = zW.locationOnScreen  //77 414
    val x = p.getInt("x", zWorldPosition.x + zWorldDimension.width + 10 )
    val y = p.getInt("y", zWorldPosition.y )
    location = Point2( x, y )

    val width = p.getInt("width", 560)
    val height = p.getInt("height", 365)
    size = (width, height)

    val c = p.get("canvas", "plane")
    c match {
      case "plane" => usePlane()
      case "sphere" => useSphere()
    }
    canvas.zoom = p.getDouble("zoom", 1)
  }

  // methods --------------------------------------------------------

  /** During dynamic map, adds another number. */
  private[calculator] def addNumberToCurve(c: Complex) = if (f != null) {
    try {
      val z = f(c)
      if(zs == null) zs = Nil
      zs = z :: zs
      updateExtremes(z)
      canvas.repaint()
    }
    //Maybe f(c) failed.
    catch {
      case _ex: Exception =>
    }
  }

  /** Adds a piclet to be mapped. */
  private[calculator] def add(piclet: Piclet) = if (f != null) {
    Complex.resetArg()
    val samples = piclet.samples //sure these are continuous?
    var zs: List[Complex] = Nil
    samples.foreach{ z =>
        try {
          val fz = f(z)
          updateExtremes(fz)
          zs = fz :: zs
        } catch {
          case _ex: Exception =>
        }
    }
    piclets = Curve(zs) :: piclets
    canvas.repaint()
  }

  private[calculator] def add( list: List[Piclet]): Unit = list foreach add

  /** during dragging in z world, the current piclet */
  private[calculator] def addCurrent(p: Piclet) =
    if (f != null) {

      //XXX maybe better:
      //this.current = f(p)

      Complex.resetArg()
      zs = Nil
      val samples = p.samples
      if ( samples != null ) {
        samples.foreach{ z =>
          try {
            val fz = f(z)
            zs = fz :: zs
          } catch {
            case _ex: Exception =>
          }
        }
      }
    canvas.repaint()
  }

  override final private[calculator] def draw(g: Graphics) = {
    if (zs != null) canvas.draw(g, zs)
    piclets.foreach{ canvas.draw(g, _) }
  }

  override private[calculator] def erase() = {
    zs = null //forget current curve
    piclets = Nil //forget all piclets
    resetExtremes()
    canvas.repaint()
  }

  /** Event listener: function has changed. */
  private[calculator] def functionChanged() = {
    erase()
    add(zW.piclets)
  }

  private[calculator] def stopDynamicMap() = {
    piclets = Curve(zs) :: piclets
    zs = null
  }
}